### PR TITLE
[TEVA-1513] Save jobs for later

### PR DESCRIPTION
**Pull request by cpjmcquillan**
## Jira ticket URL
https://dfedigital.atlassian.net/browse/TEVA-1513

## Changes in this PR
- Add jobseeker saved jobs
- Update banner link component
- Store jobseeker location
- Allow jobseeker to save jobs
- Don't add `Cache-Control` header on jobs pages
- Allow custom messages when endpoints require authentication

### CODE REVIEW

File: app/components/shared/banner_link_component.rb [hunk 1]
```diff
@@ -10,7 +10,7 @@ def initialize(icon_class:, link_id:, link_method:, link_path:, link_text:)
   end
 
   def call
-    link_to link_path, method: link_method, class: "banner-link icon icon--left icon--#{icon_class}-blue", id: link_id do
+    link_to link_path, method: link_method, class: "banner-link icon icon--left icon--#{icon_class}", id: link_id do
```
Comment (alexbowen) [inline app/components/shared/banner_link_component.rb:13]: 👍

File: app/models/jobseeker.rb [hunk 2]
```diff
@@ -1,4 +1,7 @@
 class Jobseeker < ApplicationRecord
   devise :database_authenticatable, :registerable, :recoverable, :validatable,
          :confirmable, :lockable, :trackable, :timeoutable
+
+  has_many :jobseeker_saved_vacancies, dependent: :destroy
+  has_many :saved_jobs, through: :jobseeker_saved_vacancies, source: :vacancy
```
Comment (cesidio) [inline app/models/jobseeker.rb:6]: `saved_jobs, through: :jobseeker_saved_vacancies`

Not sure but we should define just one word everywhere. Either `job` or `vacancy`. Or maybe `ad`?

File: app/controllers/jobseekers/saved_jobs_controller.rb [hunk 3]
```diff
@@ -1,3 +1,40 @@
...
+  end
+
+  def destroy
+    @saved_job.destroy
+    redirect_to_job
+  end
+
   def index; end
+
+  def params
```
Comment (cesidio) [inline app/controllers/jobseekers/saved_jobs_controller.rb:16]: This method should be private too

File: app/models/jobseeker.rb (hunk 2 above)
Comment (cpjmcquillan) [reply app/models/jobseeker.rb:6]: I had about 500 different names for the join table 😂

**Review by cesidio — APPROVED** [review]

File: app/models/jobseeker.rb (hunk 2 above)
Comment (david-mears-dfe) [reply app/models/jobseeker.rb:6]: +1 for `jobseeker_saved_jobs`

Comment (cpjmcquillan) [reply app/models/jobseeker.rb:6]: The reason for the name of the table is one I still stand by: it is a join between the jobseekers and vacancies tables. If the vacancies table is renamed I would be happy for it to change for consistency. 

The reason the relationship on the jobseekers table is called `saved_jobs` is multi-faceted (brevity, consistency with the feature and controller, and it is what it says it is - aka clearer to understand)

Comment (cpjmcquillan) [reply app/models/jobseeker.rb]: After implementing the saved jobs dashboard I retract everything I said...

The table shall be known forever more as `saved_jobs`

File: spec/support/jobseeker_helpers.rb [hunk 4]
```diff
@@ -0,0 +1,22 @@
+module JobseekerHelpers
```
Comment (cesidio) [inline spec/support/jobseeker_helpers.rb:1]: I can see a page object here

Comment (cpjmcquillan) [reply spec/support/jobseeker_helpers.rb:1]: Haha me too. I will have a go in a different PR

File: spec/system/jobseekers_can_save_a_job_spec.rb [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+      save_job
+      expect(page).to have_content(I18n.t("messages.jobseekers.saved_jobs.unauthenticated"))
+      click_on "Sign up"
+      sign_up_jobseeker
+      confirm_email_address
+      and_the_job_is_saved
+    end
+  end
+
+  def save_job
```
Comment (cesidio) [inline spec/system/jobseekers_can_save_a_job_spec.rb:78]: 👍

File: app/controllers/vacancies_controller.rb [hunk 6]
```diff
@@ -26,8 +25,6 @@ def show
     @similar_jobs = Search::SimilarJobs.new(vacancy).similar_jobs
 
     VacancyPageView.new(vacancy).track unless publisher_signed_in? || smoke_test?
-
-    expires_in 5.minutes, public: true
```
Comment (cesidio) [inline app/controllers/vacancies_controller.rb:29]: we should monitor the performance of this choice

**Review by cesidio — APPROVED** [review]

**Pull request merged**